### DISCUSSION TITLE
Let's list all upgrades you want to see.

### DISCUSSION THREAD

**Discussion by CodingTheUnknown** [jamesmccannon02/OctoPrint-Tplinkautoshutdown#13] [Ideas]
Please write all upgrades you want to see here. Helps to keep them all in one place.

**Comment by madnlooney** [comment]
i use the KP115 with energy monitoring, maybe having that info available with the cost per kw too.

**Reply by CodingTheUnknown** [reply]
Currently looking at how I'm going to this

**Reply by CodingTheUnknown** [reply]
@madnlooney If i write the software for this, would you be willing to test it for me? I only have a KP100 so can't test this code. I will distribute it in a way that you can download the plugin using the URL method within the plugin manager so that i don't disrupt the stable releases.  All you will need to do is install it and test it with a few tests i create. You would however have to temporarily uninstall the stable version.

**Reply by madnlooney** [reply]
Yeah I can test it for you. I use the other tplink plugin currently but happy to test your plugin

**Reply by CodingTheUnknown** [reply]
@madnlooney [https://github.com/jamesmccannon02/OctoPrint-Tplinkautoshutdown/archive/develop-powerStrip.zip](https://github.com/jamesmccannon02/OctoPrint-Tplinkautoshutdown/archive/develop-powerStrip.zip)

This is the link to test the KP303 support. You will need to uninstall the current version and then copy and paste the link above into the install manager URL option

**Reply by madnlooney** [reply]
Ive just seen this is running on python 3. ill need to look at upgrading from 2 then i can test this. also i have a kp115

**Reply by CodingTheUnknown** [reply]
If you will be willing to that would be great if you could. Thanks. I know that there is a bug that when you click update connection, you will need to click it again for the information within the preference window to populate but looking to fix that.

**Reply by madnlooney** [reply]
its the way plugins are going to be build on the newer version then its only a matter of backup octoprint settings and then reflash and restore. i cant use the script as im on 0.16

**Reply by CodingTheUnknown** [reply]
Yes from what i am ware. Haven't done it personally though

**Reply by madnlooney** [reply]
I've updated everything. installed your plugin. just need to test the auto off function.
Is there any chance of power monitoring information to be shown on the KP115

**Reply by madnlooney** [reply]
@jamesmccannon02 

It power off after print finished. Have all set up with both check boxes marked. The off button on the banner works so I know the plugin sees the plug

**Reply by CodingTheUnknown** [reply]
Great thanks for the feedback.

**Comment by RodLaird** [comment]
I am using KP303's - and the octoprint server is one of the plugs on the printer strip. So the shutdown also shuts down it - which is fine actually but not ideal.

Also - being able to specify a delay before shutdown would be nice. Indeed a "shutdown after idle" time - so that if there were sequential prints that I was doing during the day, and then one that finished overnight, the printer would auto shutdown after idling...

The Kasa python modules would seem to be able to deal with the strip plug without much of a problem...

**Reply by CodingTheUnknown** [reply]
I'll have a look at implementing this. Would you be willing to test it if i program this?

**Reply by CodingTheUnknown** [reply]
Can i confirm, are you currently using this at the moment, just released an update and would appreciate to know if it is working. Can only test so much in my env.

**Reply by RodLaird** [reply]
0.2.0 seems to work fine!

**Reply by CodingTheUnknown** [reply]
You might be interested in 0.2.1. It now has a feature to allow the OctoPrint timelapse render to finish before the plug is turned off.

**Reply by CodingTheUnknown** [reply]
@RodLaird If you wouldn't mind, please can you test the support for the KP303. I have developed a new version. You will need to uninstall the current TpHandler and then install the plugin using the URL

[https://github.com/jamesmccannon02/OctoPrint-Tplinkautoshutdown/archive/develop-powerStrip.zip](https://github.com/jamesmccannon02/OctoPrint-Tplinkautoshutdown/archive/develop-powerStrip.zip)

**Comment by RodLaird** [comment]
Of course – no sweat. I use the Kasa module in some of my other household automation Python code and am pretty sure you will find no challenges. 
 
Rod Laird
+61 4 1816 6105

**Reply by CodingTheUnknown** [reply]
If you wouldn't mind, please can you test the support for the KP303. I have developed a new version. You will need to uninstall the current TpHandler and then install the plugin using the URL

[https://github.com/jamesmccannon02/OctoPrint-Tplinkautoshutdown/archive/develop-powerStrip.zip](https://github.com/jamesmccannon02/OctoPrint-Tplinkautoshutdown/archive/develop-powerStrip.zip)

**Comment by escfoe2** [comment]
I do not enjoy the messages that pop up when pressing the buttons.  Please put the confirmation messages within the buttons themselves.

First click > Button text changes to "Confirm? (Secs_Remaining_Before_TimeOut)"
Second click > If clicked before timeout > Execute Button Action

The messages that pop up are not intuitive.

**Reply by CodingTheUnknown** [reply]
Will look into doing this. Nice idea. Thanks

**Reply by escfoe2** [reply]
Did you do it yet? lol

**Reply by escfoe2** [reply]
I'm sure this isn't Visual Basic but in VB we'd do it something like this:
`
Private WaitingForOnClick as Boolean = False
Sub btnOn_Click(sender as Object, e as EventArgs)

    If btnOn.Text = "Turn On!" Then
        WaitingForOnClick = True
        Task.Factory.StartNew(Sub()
               For X as Integer = 1 to 3
                   If Not WaitingForOnClick Then Exit Sub
                   btnOn.Text = "Confirm (" & 3 - (X - 1)
                   Thread.Sleep(1000)
               Next

               btnOn.Text = "Turn On!"
               WaitingForOnClick = False
           End Sub)
    Else
        'PERFORM DEVICE TURN ON
        WaitingForOnClick = False
        btnOn.Text = "Turn On!"
    End If

End Sub
`

**Reply by CodingTheUnknown** [reply]
I am currently looking into this but about to release support for powerStrips so that has taken my attention this week. However, my issue is that i need to look how to bring the flask side of the server together with javascript and all at the same time as not holding up async processes otherwise my application goes against the publication rules of OctoPrint

**Comment by escfoe2** [comment]
Also, I would like a little padding added to the right of the 'On' button.  The same padding that's in-between the buttons will suffice.
![TpLink](https://user-images.githubusercontent.com/17956005/108634234-adf55000-7446-11eb-8037-0f91d17cdd5e.png)

**Reply by CodingTheUnknown** [reply]
Will add some padding. Thanks for the feedback as I hadn't thought of this issue. Will fix in 0.2.2. Will aim to have it done by Tuesday.

**Reply by CodingTheUnknown** [reply]
@escfoe2 Just released 0.2.2 with a fix to this. It will add 10px margin. Let me know what you think as version 1.0.0 will come soon and i can make further changes then

**Comment by RodLaird** [comment]
Hmmm. Turns out the plugin does turn off printer manually but not 
automatically with the KP303. Looks like need to wait for specific code. 
Spaghetti detective also not completing with plugin operative some some 
weird interactions.

cheers


Rod Laird
+61 418 166 105

**Reply by CodingTheUnknown** [reply]
Currently looking how this can be done with the current code i have. The difficulty i have is not the backend but i have very little front end experience so need to figure a way to make it user friendly enough with multiple device options. 


I'll look to see what options i can setup for you

**Reply by CodingTheUnknown** [reply]
@RodLaird I believe i now have a way to add support for smart strips for up to 3 plug socket. I'll contact you when it is complete for you to test.

**Reply by CodingTheUnknown** [reply]
If you wouldn't mind, please can you test the support for the KP303. I have developed a new version. You will need to uninstall the current TpHandler and then install the plugin using the URL

[https://github.com/jamesmccannon02/OctoPrint-Tplinkautoshutdown/archive/develop-powerStrip.zip](https://github.com/jamesmccannon02/OctoPrint-Tplinkautoshutdown/archive/develop-powerStrip.zip)

**Reply by CodingTheUnknown** [reply]
Hi, don't know if you have tried using the plugin over the past 3 days, but there is a version you can download using the link above that i have had confirmation works. If you would like to give it a go that would be great. Please note when you have downloaded it. Placed your ip into the box within the settings, you will need to re-start the server before it will work.

**Reply by RodLaird** [reply]
Hi James,

This now seems to work fine... Congratulations. I seem to remember the Kasa API allows pulling down the user assigned switch names. Is it possible to pull them into the interface? I was concerned that The Spaghetti Detective plugin might not get the signals it needs if your plugin turned off the power to soon - but it seems OK! A user selectable delay might be nice though.

regards

**Comment by RodLaird** [comment]
Yup; I understand the difficulties!

cheers


Rod Laird
+61 418 166 105

**Comment by RodLaird** [comment]
I certainly will; this week in Tasmania... home Monday

cheers


Rod Laird
+61 418 166 105

**Comment by RodLaird** [comment]
Will do!

cheers


Rod Laird
+61 418 166 105

**Comment by WashUDave** [comment]
Requesting an "Auto-On' option for smart plugs. I use it with my light to record time-lapses, and sometimes forget I need to manually turn it on.

Thanks!

**Comment by pabutterworth** [comment]
I love the plugn, thanks

It would be great if it attempted to auto-connect to the printer after a number of seconds. 

cheers
Paul

**Reply by CodingTheUnknown** [reply]
> I love the plugn, thanks
> 
> 
> 
> It would be great if it attempted to auto-connect to the printer after a number of seconds. 
> 
> 
> 
> cheers
> 
> Paul

This was tried but for some reason it caused more problems than it solved so was removed as a feature

**Comment by pabutterworth** [comment]
Oh ok, well thanks for trying! And thanks for responding so quickly.
Cheers

On Mon, 7 Feb 2022 at 16:51, jamesmccannon02 ***@***.***>
wrote:

> I love the plugn, thanks
>
> It would be great if it attempted to auto-connect to the printer after a
> number of seconds.
>
> cheers
>
> Paul
>
> This was tried but for some reason it caused more problems than it solved
> so was removed as a feature
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/jamesmccannon02/OctoPrint-Tplinkautoshutdown/discussions/13#discussioncomment-2126988>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AD737VYDWBRX7OEEDSKJQK3UZ72CHANCNFSM4X3GHWYA>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
> You are receiving this because you commented.Message ID:
> <jamesmccannon02/OctoPrint-Tplinkautoshutdown/repo-discussions/13/comments/2126988
> @github.com>
>
-- 
Paul Butterworth
+44 7768 252250
